### DISCUSSION TITLE
What to do when migration doesn't solve table mismatch

### DISCUSSION THREAD

**Discussion by viperciyes** [serverpod/serverpod#2797] [Q&A] [answered]
I did some changes to an existing table definition, run serverpod generate, create migration etc. and applied it, but table definitions wouldN't update. then I created a repair migration and it didn't solve as well. What can I do now?
```

WARNING: The database does not match the target database:
 - Table "users" is not like the target database.
Hint: Did you forget to apply the migrations (--apply-migrations) or run a repair migration (--apply-repair-migration)?

```

**Comment by SandPod** [comment]
Hey @viperciyes 👋 

We are working on improving the developer experience when database mismatches happen, hopefully it will be included in the next relase, issue here: https://github.com/serverpod/serverpod/issues/1262, PR here: https://github.com/serverpod/serverpod/pull/2778

Just to get a better understanding of how you ended up in this state, did you do any manual modifications to the database or did you only apply changes through Serverpod models?

Which version of Serverpod are you using?

**Reply by viperciyes** [reply]
hey there, sorry for the late reply. 
I only did changes to models and applied migrations. I wrote a few rows through the app I'm building, and later on I might have truncated the table, not really sure about that part, but I know I didn't add or remove any columns. 
I decided to change some things in the model file. then run the commands to generate and create migration and applied it. it didn't work, I forced it, then it gave this warning and I created a repair migration, and applied it and it didn't solve it either.
Something I noticed: In the generated code files, I found reference to the model field I deleted from the model. So I deleted the generated models and when I run serverpod generate, that field still comes back somehow.

**Reply by SandPod** [reply]
May I ask what OS you are running this on?

We have a reported issue here that might be somewhat related to your case, where the runtime of the server instance is not the same as the one you are seeing in the file explorer: https://github.com/serverpod/serverpod/issues/2857

**Reply by viperciyes** [reply]
I'm on Mac running macOs Monterey, server is on Docker & working on localhost at the moment.

**Reply by SandPod** [reply]
Hmm, most of the team is also running on MacOS so I don't expect this type of issue to be present there.

**Reply by viperciyes** [reply]
If there is anything I can do to help detect the problem, please tell me. otherwise, I'll do what other people did I guess, create a new project, copy models and endpoints, generate code and then copy everything back here.

**Reply by SandPod** [reply]
Does the problem persist if you "move" the project? Could you scrape all sensitive content from the project and host it in a public github repo so that I could take a look at it?

**Reply by viperciyes** [reply]
Sorry for the late reply. after a long time I made a mess of the project today and was able to put it together finally. and even after a fresh project setup, code copy paste to regenerate, the issue is there (I didn't start a new db though). One thing: during the code copying, the editor complained about not being able to find module auth, but now it doesn't. this is where I have the problem (table users), so maybe it is related.
Here is the repo: [stokla](https://github.com/viperciyes/stokla_p)

**Reply by SandPod** [reply, marked as answer]
Thanks for getting back to us.

After you added the auth module, did you remember to do run `pub update` before creating the migrations?

If the migration is created before you have fetched any Serverpod module, our database migration tool will not be able to determine what needs to be created in order to satisfy the database requirements of the module, and therefore not create tables required by the module.

**Reply by viperciyes** [reply]
well, I certainly did not run `pub update`. that might be the reason. I did not add more tables but changed some fields.

**Reply by SandPod** [reply]
Then I would suspect that is the culprit in this particular case :)

**Reply by viperciyes** [reply]
I checked the documentation but I couldn't find anything about a pub update after adding a module. Either I couldn't find it, or the docs need another paragraph. Thanks anyways!

**Reply by SandPod** [reply]
Aah, that is our bad. Will try and clarify this in the documentation.

**Reply by viperciyes** [reply]
I can confirm that after `pub update`, the warning is now gone.
